src/OrchardCore/OrchardCore.Environment.Shell.Abstractions/Builders/ShellContext.cs

```diff
@@ -143,10 +173,51 @@ public ServiceScopeWrapper(IServiceScope serviceScope)
 
             public IServiceProvider ServiceProvider { get; }
 
+            /// <summary>
+            /// Returns true is the shell context should be disposed consequently to this scope being released.
+            /// </summary>
+            private bool ScopeReleased()
+            {
+                var refCount = Interlocked.Decrement(ref _shellContext._refCount);
+
+                if (_shellContext._released && refCount == 0)
+                {
```

Comment:
Hmm, maybe this is here that we could use `CanTerminate()` which also check `_disposed`.

Reply:
No, I want to use `refCount` and not `_refCount`

Reply:
Ah okay.
In my case i was using a `_shellContext.ScopeEnded()` so that the shell was disposing itself.
But you need to do things here in the wrapper to be able to resolve `IModularTenantEvents`.
So, that's ok.